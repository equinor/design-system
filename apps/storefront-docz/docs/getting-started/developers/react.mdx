---
title: Developers
tabs:
  - Introduction
  - CSS framework
  - React
  - Figma
mode: draft
route: /getting-started/developers/react/
---

:warning: Information coming soon
