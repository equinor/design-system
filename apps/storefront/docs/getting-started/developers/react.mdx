---
title: Developers
metaTitle: Developers react
tabs:
  - Introduction
  - CSS framework
  - React
  - Figma
mode: draft
route: /getting-started/developers/react/
---

:warning: Information coming soon
