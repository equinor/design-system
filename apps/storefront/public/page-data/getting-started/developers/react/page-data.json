{"componentChunkName":"component---docs-getting-started-developers-react-mdx","path":"/getting-started/developers/react/","result":{"pageContext":{"frontmatter":{"title":"Developers","metaTitle":"Developers react","tabs":["Introduction","CSS framework","React","Figma"],"mode":"draft","route":"/getting-started/developers/react/"},"entry":{"id":"6f9784da00517d57e4627269567073e6","filepath":"docs/getting-started/developers/react.mdx","fullpath":"/Users/vnys/Kode/storefront-replace/apps/storefront/docs/getting-started/developers/react.mdx","route":"/getting-started/developers/react/","slug":"docs-getting-started-developers-react","name":"React","menu":"","headings":[]}}}}