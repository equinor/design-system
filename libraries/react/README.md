@eds-ui/react
===============

1. `npm i -g pnpm`
1. `pnpm install`
1. `pnpm recursive install`

## Run storybook

`pnpm start`

## Build modules

`pnpm recursive run build`
