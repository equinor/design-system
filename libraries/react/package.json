{
  "name": "@eds-ui/react",
  "private": true,
  "version": "0.0.1",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "jest",
    "storybook": "start-storybook"
  },
  "keywords": [],
  "author": "",
  "license": "AGPL",
  "dependencies": {
    "react": "^16.8.6",
    "react-dom": "^16.8.6"
  },
  "devDependencies": {
    "@babel/cli": "^7.4.4",
    "@babel/core": "^7.4.4",
    "@babel/preset-env": "^7.4.4",
    "@babel/preset-react": "^7.0.0",
    "@storybook/react": "^5.0.11",
    "babel-plugin-styled-components": "^1.10.0",
    "enzyme": "^3.9.0",
    "enzyme-adapter-react-16": "^1.12.1",
    "jest": "^24.8.0",
    "jest-styled-components": "^6.3.1",
    "rollup": "^1.11.3",
    "rollup-plugin-babel": "^4.3.2",
    "rollup-plugin-commonjs": "^9.3.4",
    "rollup-plugin-node-resolve": "^4.2.3",
    "styled-components": "^4.2.0"
  },
  "peerDependencies": {
    "react": "^16.8.6",
    "react-dom": "^16.8.6",
    "styled-components": "^4.2.0",
    "prop-types": "^15.7.2"
  },
  "jest": {
    "verbose": true,
    "setupFilesAfterEnv": [
      "<rootDir>/testSetup.js"
    ],
    "testPathIgnorePatterns": [
      "<rootDir>/(node_modules)"
    ],
    "transformIgnorePatterns": [
      "<rootDir>/(node_modules)"
    ],
    "haste": {
      "providesModuleNodeModules": [
        ".*"
      ]
    },
    "roots": [
      "packages/node_modules"
    ]
  }
}
