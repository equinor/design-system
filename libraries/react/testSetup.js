import React from 'react'
import Enzyme from 'enzyme'
import 'jest-styled-components'
import styled from 'styled-components'
import Adapter from 'enzyme-adapter-react-16'

Enzyme.configure({ adapter: new Adapter() })
