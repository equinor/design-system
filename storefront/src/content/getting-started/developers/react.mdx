---
title: React
tabs:
    - Introduction
    - CSS framework
    - React
    - Figma
    - Contributing
mode: draft
---


:warning: Information coming soon
