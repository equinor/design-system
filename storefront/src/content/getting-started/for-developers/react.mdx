---
title: React
tabs:
    - Introduction
    - CSS framework
    - React
    - Figma
    - Contributing
---

Developers React
