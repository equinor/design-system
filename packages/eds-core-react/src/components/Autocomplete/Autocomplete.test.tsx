import { useState } from 'react'
import {
  render,
  fireEvent,
  screen,
  within,
  waitFor,
} from '@testing-library/react'
import styled from 'styled-components'
import { Autocomplete } from '.'

const itemObjects = [{ label: 'One' }, { label: 'Two' }, { label: 'Three' }]
const items = ['One', 'Two', 'Three']
const labelText = 'Select label test'

const mockResizeObserver = jest.fn(() => ({
  observe: jest.fn(),
  disconnect: jest.fn(),
  unobserve: jest.fn(),
}))

beforeAll(() => {
  window.ResizeObserver = mockResizeObserver

  //https://github.com/TanStack/virtual/issues/641
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  //@ts-ignore
  Element.prototype.getBoundingClientRect = jest.fn(() => {
    return {
      width: 120,
      height: 120,
      top: 0,
      left: 0,
      bottom: 0,
      right: 0,
    }
  })
})

describe('Autocomplete', () => {
  it('Matches snapshot', async () => {
    render(<Autocomplete options={items} label={labelText} />)

    const autocomplete = screen.getAllByLabelText(labelText)
    const input = autocomplete[0]

    fireEvent.click(input)

    const openAutocomplete = await screen.findAllByLabelText(labelText)
    const optionsList = openAutocomplete[1]

    expect(optionsList).toMatchSnapshot()
  })
  it('Has provided label', async () => {
    render(<Autocomplete label={labelText} options={items} />)

    // The same label is used for both the input field and the list of options
    const labeledNodes = await screen.findAllByLabelText(labelText)
    const input = labeledNodes[0]
    const optionsList = labeledNodes[1]

    expect(input).toBeDefined()
    expect(input).toHaveAccessibleName(labelText)
    expect(input.nodeName).toBe('INPUT')

    expect(optionsList).toBeDefined()
    expect(optionsList).toHaveAccessibleName(labelText)
    expect(optionsList.nodeName).toBe('UL')
  })

  it('Has provided ReactNode label', async () => {
    render(<Autocomplete label={<div>{labelText}</div>} options={items} />)

    // The same label is used for both the input field and the list of options
    const labeledNodes = await screen.findAllByLabelText(labelText)
    const input = labeledNodes[0]
    const optionsList = labeledNodes[1]

    expect(input).toBeDefined()
    expect(input).toHaveAccessibleName(labelText)
    expect(input.nodeName).toBe('INPUT')

    expect(optionsList).toBeDefined()
    expect(optionsList).toHaveAccessibleName(labelText)
    expect(optionsList.nodeName).toBe('UL')
  })

  it('Has provided option label', async () => {
    const labler = (text: string) => `${text}+1`
    render(
      <Autocomplete
        options={itemObjects}
        label={labelText}
        optionLabel={(item) => labler(item.label)}
      />,
    )
    const labeledNodes = await screen.findAllByLabelText(labelText)
    const optionsList = labeledNodes[1]

    const buttonNode = await screen.findByLabelText('toggle options', {
      selector: 'button',
    })
    expect(optionsList.childNodes).toHaveLength(0)

    fireEvent.click(buttonNode)
    expect(await within(optionsList).findAllByRole('option')).toHaveLength(3)

    const options = await within(optionsList).findAllByRole('option')
    expect(within(options[0]).getByText(labler(items[0]))).toBeDefined()
    expect(within(options[1]).getByText(labler(items[1]))).toBeDefined()
    expect(within(options[2]).getByText(labler(items[2]))).toBeDefined()
  })

  it('Can render custom items with optionComponent', async () => {
    type Item = {
      label: string
    }
    function CustomItem(option: Item) {
      const { label } = option
      return <h1>{label}</h1>
    }
    render(
      <Autocomplete
        options={itemObjects}
        label={labelText}
        optionLabel={(item) => item.label}
        optionComponent={CustomItem}
      />,
    )
    const labeledNodes = await screen.findAllByLabelText(labelText)
    const optionsList = labeledNodes[1]

    const buttonNode = await screen.findByLabelText('toggle options', {
      selector: 'button',
    })
    expect(optionsList.childNodes).toHaveLength(0)

    fireEvent.click(buttonNode)
    expect(await within(optionsList).findAllByRole('option')).toHaveLength(3)

    const options = await within(optionsList).findAllByRole('option')
    expect(within(options[0]).getByText(items[0])).toBeDefined()
    const heading = screen.getByText(items[0])
    expect(heading.nodeName).toBe('H1')
  })

  it('Can be disabled', async () => {
    render(<Autocomplete label={labelText} options={items} disabled />)
    const labeledNodes = await screen.findAllByLabelText(labelText)
    const input = labeledNodes[0]

    expect(input).toBeDisabled()
  })

  it('Can preselect specific options', async () => {
    render(
      <Autocomplete
        options={items}
        label={labelText}
        initialSelectedOptions={['One', 'Two']}
        multiple
      />,
    )
    const input = await screen.findByPlaceholderText('2/3 selected')
    fireEvent.click(input)

    const checkboxes = await screen.findAllByRole('checkbox')
    const checked = checkboxes.filter((x) => x.hasAttribute('checked'))

    expect(input).toBeDefined()
    expect(checked.length).toBe(2)
  })

  it('Can select all options', async () => {
    const onChange = jest.fn()
    render(
      <StyledAutocomplete
        //a bug in styled-components 6.1.8 breaks the conditional type for optionLabel when using styled(Autocomplete)
        optionLabel={(label: string) => label}
        label={labelText}
        options={items}
        data-testid="styled-autocomplete"
        multiple={true}
        allowSelectAll={true}
        onOptionsChange={onChange}
      />,
    )

    const labeledNodes = await screen.findAllByLabelText(labelText)
    const optionsList = labeledNodes[1]

    const buttonNode = await screen.findByLabelText('toggle options', {
      selector: 'button',
    })

    fireEvent.click(buttonNode)

    const options = await within(optionsList).findAllByRole('option')
    fireEvent.click(options[0])

    await waitFor(() => {
      expect(onChange).toHaveBeenCalledWith({ selectedItems: items })
    })

    fireEvent.click(options[0])

    await waitFor(() => {
      expect(onChange).toHaveBeenCalledWith({ selectedItems: [] })
    })
  })

  it('Can deselect complex options', async () => {
    const onChange = jest.fn()
    const opts = [
      { label: 'Its', value: 'relationship' },
      { label: 'Complicated', value: 'status' },
    ]
    render(
      <Autocomplete
        optionLabel={(o) => o.label}
        itemCompare={(o1, o2) => o1.value === o2.value}
        label={labelText}
        options={opts}
        data-testid="styled-autocomplete"
        multiple={true}
        onOptionsChange={onChange}
        selectedOptions={[
          {
            label: 'Its',
            value: 'relationship',
          },
        ]}
      />,
    )

    const labeledNodes = await screen.findAllByLabelText(labelText)
    const optionsList = labeledNodes[1]

    const buttonNode = await screen.findByLabelText('toggle options', {
      selector: 'button',
    })

    fireEvent.click(buttonNode)

    const options = await within(optionsList).findAllByRole('option')
    fireEvent.click(options[0])

    await waitFor(() => {
      expect(onChange).toHaveBeenCalledWith({ selectedItems: [] })
    })
  })

  it('Can open the options on button click', async () => {
    render(<Autocomplete options={items} label={labelText} />)

    const labeledNodes = await screen.findAllByLabelText(labelText)
    const optionsList = labeledNodes[1]

    const buttonNode = await screen.findByLabelText('toggle options', {
      selector: 'button',
    })
    expect(optionsList.childNodes).toHaveLength(0)

    fireEvent.click(buttonNode)

    expect(await within(optionsList).findAllByRole('option')).toHaveLength(3)
  })

  type ControlledProps = {
    onOptionsChange: () => void
  }

  const ControlledAutoComplete = ({ onOptionsChange }: ControlledProps) => {
    const [selected, setSelected] = useState([])
    return (
      <Autocomplete
        multiple
        options={items}
        label={labelText}
        selectedOptions={selected}
        onOptionsChange={(changes) => {
          setSelected(changes.selectedItems)
          onOptionsChange()
        }}
      />
    )
  }

  it('Can be a controlled component', async () => {
    const handleChange = jest.fn()
    render(<ControlledAutoComplete onOptionsChange={handleChange} />)
    const labeledNodes = await screen.findAllByLabelText(labelText)
    const optionsList = labeledNodes[1]
    const buttonNode = await screen.findByLabelText('toggle options', {
      selector: 'button',
    })

    await waitFor(() => {
      expect(handleChange).toHaveBeenCalledTimes(0)
    })
    fireEvent.click(buttonNode)
    const options = await within(optionsList).findAllByRole('option')
    fireEvent.click(options[2])

    await waitFor(() => {
      expect(handleChange).toHaveBeenCalledTimes(1)
    })
  })

  it('Can filter results by input value', async () => {
    render(<Autocomplete options={items} label={labelText} />)
    const labeledNodes = await screen.findAllByLabelText(labelText)
    const input = labeledNodes[0]
    const optionsList = labeledNodes[1]

    const buttonNode = await screen.findByLabelText('toggle options', {
      selector: 'button',
    })
    expect(optionsList.childNodes).toHaveLength(0)

    fireEvent.click(buttonNode)
    expect(await within(optionsList).findAllByRole('option')).toHaveLength(3)

    fireEvent.change(input, {
      target: { value: 'ree' },
    })
    const filteredOptions = await within(optionsList).findAllByRole('option')
    expect(filteredOptions).toHaveLength(1)
    expect(within(filteredOptions[0]).getByText('Three')).toBeDefined()
  })

  it('Second option is first when first option is disabled', async () => {
    render(
      <Autocomplete
        options={items}
        label={labelText}
        optionDisabled={(item) => item === items[0]}
      />,
    )

    const labeledNodes = await screen.findAllByLabelText(labelText)
    const input = labeledNodes[0]
    const optionsList = labeledNodes[1]

    fireEvent.keyDown(input, { key: 'ArrowDown' })
    const options = await within(optionsList).findAllByRole('option')
    expect(options).toHaveLength(2) // since one option is disabled
    expect(await within(options[0]).findByText(items[1])).toBeDefined()

    const withDisabledOptions = await within(optionsList).findAllByRole(
      'option',
      {
        hidden: true,
      },
    )
    expect(withDisabledOptions[0]).toHaveAttribute('aria-hidden')
    expect(
      await within(withDisabledOptions[0]).findByText(items[0]),
    ).toBeDefined()
  })

  it('Clears the input text on blur when no option is selected', async () => {
    render(<Autocomplete options={items} label={labelText} />)
    const labeledNodes = await screen.findAllByLabelText(labelText)
    const input = labeledNodes[0]

    fireEvent.change(input, {
      target: { value: 'ree' },
    })

    fireEvent.blur(input)
    expect(input).toHaveValue('')
  })

  it('Correctly handles keypresses up/down when all options are disabled', async () => {
    render(
      <Autocomplete
        options={items}
        label={labelText}
        // Somewhat contrived condition to emulate a scenario where an undefined item would return true for the 'option' being disabled
        optionDisabled={(item) => item !== items[0]}
      />,
    )
    const labeledNodes = await screen.findAllByLabelText(labelText)
    const input = labeledNodes[0]
    const optionsList = labeledNodes[1]

    fireEvent.keyDown(input, { key: 'ArrowDown' })
    const options = await within(optionsList).findAllByRole('option')
    expect(options).toHaveLength(1) // since all but one options are disabled

    fireEvent.change(input, {
      target: { value: 'asfsggsdhfj' },
    })
    const optionsAfterSearch = within(optionsList).queryAllByRole('option')
    expect(optionsAfterSearch).toHaveLength(0) // since all are filtered out

    // Prevent regression: key up/down when options are disabled causes infinite loop
    fireEvent.keyDown(input, { key: 'ArrowDown' })
    fireEvent.blur(input)
    expect(input).toHaveValue('')
  })

  const StyledAutocomplete = styled(Autocomplete)`
    clip-path: unset;
  `

  it('Can extend the css for the component & props are passed correctly to input', async () => {
    const { container } = render(
      <StyledAutocomplete
        //a bug in styled-components 6.1.8 breaks the conditional type for optionLabel when using styled(Autocomplete)
        optionLabel={(label: string) => label}
        label="test"
        options={items}
        data-testid="styled-autocomplete"
      />,
    )

    const autocomplete = await screen.findByTestId('styled-autocomplete')

    // eslint-disable-next-line testing-library/no-node-access
    expect(container.firstChild).toHaveStyleRule('clip-path', 'unset')
    expect(autocomplete.nodeName).toBe('INPUT')
  })
})
