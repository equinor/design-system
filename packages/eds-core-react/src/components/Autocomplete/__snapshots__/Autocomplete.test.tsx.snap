// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete Matches snapshot 1`] = `
.c0 {
  margin: 0;
  color: var(--eds_paragraph__body_short_color, rgba(61, 61, 61, 1));
  font-family: Equinor;
  font-size: 1.000rem;
  font-weight: 400;
  line-height: 1.250em;
  text-align: left;
}

.c2 {
  display: flex;
  grid-area: 1/-1;
  align-items: center;
  align-self: start;
  margin: 0;
  list-style: none;
  background-color: var(--eds_ui_background__default, rgba(255, 255, 255, 1));
  user-select: none;
  overflow: hidden;
  touch-action: manipulation;
  z-index: 3;
  cursor: default;
  margin: 0;
  color: var(--eds_text__static_icons__default, rgba(61, 61, 61, 1));
  font-family: Equinor;
  font-size: 1.000rem;
  font-weight: 400;
  line-height: 1.000em;
  letter-spacing: 0.013em;
  text-align: left;
  padding-left: 24px;
  padding-top: 0;
  padding-right: 24px;
  padding-bottom: 0;
}

.c3 {
  padding-left: 0;
  padding-top: 16px;
  padding-right: 0;
  padding-bottom: 16px;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
  overflow-wrap: anywhere;
}

.c1 {
  background-color: var(--eds_ui_background__default, rgba(255, 255, 255, 1));
  box-shadow: 0 7px 8px rgba(0, 0, 0, 0.2),0 5px 22px rgba(0, 0, 0, 0.12),0 12px 17px rgba(0, 0, 0, 0.14);
  border-radius: 4px;
  overflow-y: auto;
  overflow-x: hidden;
  padding: 0;
  display: grid;
}

@media (hover:hover) and (pointer:fine) {

}

@media (hover:hover) and (pointer:fine) {

}

@supports (-moz-appearance:none) {
  .c3 {
    overflow: clip;
  }
}

@supports (-moz-appearance:none) {
  .c1 {
    scrollbar-width: thin;
  }
}

<ul
  aria-labelledby="downshift-:r0:-label"
  class="c0 c1"
  id="downshift-:r0:-menu"
  role="listbox"
  style="max-height: 300px;"
>
  <li
    role="presentation"
    style="height: 144px; margin: 0px; grid-area: 1 / -1;"
  />
  <li
    aria-disabled="false"
    aria-hidden="false"
    aria-posinset="1"
    aria-selected="false"
    aria-setsize="3"
    class="c2"
    data-index="0"
    id="downshift-:r0:-item-0"
    role="option"
    style="transform: translateY(0px); height: 48px;"
  >
    <span
      class="c3"
    >
      One
    </span>
  </li>
  <li
    aria-disabled="false"
    aria-hidden="false"
    aria-posinset="2"
    aria-selected="false"
    aria-setsize="3"
    class="c2"
    data-index="1"
    id="downshift-:r0:-item-1"
    role="option"
    style="transform: translateY(48px); height: 48px;"
  >
    <span
      class="c3"
    >
      Two
    </span>
  </li>
  <li
    aria-disabled="false"
    aria-hidden="false"
    aria-posinset="3"
    aria-selected="false"
    aria-setsize="3"
    class="c2"
    data-index="2"
    id="downshift-:r0:-item-2"
    role="option"
    style="transform: translateY(96px); height: 48px;"
  >
    <span
      class="c3"
    >
      Three
    </span>
  </li>
</ul>
`;
